topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java

```diff
@@ -274,4 +282,95 @@ public ActorFuture<ClusterTopology> initialize() {
       }
     }
   }
+
+  /** Initializer that allows rolling update from 8.3.x to 8.4.x */
+  class RollingUpdateAwareInitializerV83ToV84 implements TopologyInitializer {
+    private static final Logger LOGGER =
+        LoggerFactory.getLogger(RollingUpdateAwareInitializerV83ToV84.class);
+    private static final Duration RETRY_DELAY = Duration.ofMillis(200);
+    private final ClusterMembershipService membershipService;
+    private final StaticInitializer staticInitializer;
+    private final int staticClusterSize;
+    private final ConcurrencyControl executor;
+    private final CompletableActorFuture<ClusterTopology> initializeFuture;
+
+    public RollingUpdateAwareInitializerV83ToV84(
+        final ClusterMembershipService membershipService,
+        final StaticConfiguration staticConfiguration,
+        final ConcurrencyControl executor) {
+      this.membershipService = membershipService;
+      staticInitializer = new StaticInitializer(staticConfiguration);
+      staticClusterSize = staticConfiguration.clusterMembers().size();
+      this.executor = executor;
+      initializeFuture = new CompletableActorFuture<>();
+    }
+
+    @Override
+    public ActorFuture<ClusterTopology> initialize() {
+      if (staticClusterSize == 1) {
+        // Cluster will be initialized via normal static initializer
+        initializeFuture.complete(ClusterTopology.uninitialized());
+        return initializeFuture;
+      }
+
+      final Version version = membershipService.getLocalMember().version();
+      if (isVersion8_4(version)) {
+        final boolean knowOtherMembers = hasOtherReachableMembers(membershipService);
+        if (knowOtherMembers && isRollingUpdate(membershipService)) {
+          LOGGER.debug(
+              "Cluster is doing rolling update. Cannot initialize cluster topology via gossip. Initializing cluster topology from static configuration.");
+          staticInitializer.initialize().onComplete(initializeFuture);
+        } else if (!knowOtherMembers) {
+          LOGGER.debug(
+              "No other members are reachable. Cannot initialize topology. Will retry in {}",
+              RETRY_DELAY);
+          executor.schedule(RETRY_DELAY, this::initialize);
+        } else {
+          // do not initialize. It is not rolling update so initialize via gossip or sync
+          initializeFuture.complete(ClusterTopology.uninitialized());
+        }
+      } else {
+        initializeFuture.complete(ClusterTopology.uninitialized());
+      }
+      return initializeFuture;
+    }
+
+    private boolean hasOtherReachableMembers(final ClusterMembershipService membershipService) {
+      return membershipService.getMembers().stream()
+          .map(Member::id)
+          .anyMatch(memberId -> !memberId.equals(membershipService.getLocalMember().id()));
+    }
+
+    private boolean isRollingUpdate(final ClusterMembershipService membershipService) {
```

Comment:
:thought_balloon: Don't we need to make sure that we know about enough members? If I recall correctly, membership service includes standalone gateways. So it could happen that someone updates gateways first, then does a rolling update of brokers. The first broker could be unlucky, only find a gateway in version 8.4.0 and then assume it's _not_ a rolling update.
I think we need to at least exclude gateways here.

Reply:
:+1: Excluding gateways is required. I will update this. If it finds only one other broker at 8.4.x and assumes it is not a rolling update it is fine, because it can still get the initialized topology from that one broker.

Reply:
I added a naive way to determine if a member is a broker by checking if the id is an integer. :see_no_evil:

Reply:
:+1: I think that's fine, it's a workaround either way :sweat_smile: